### PR TITLE
WIP: Windows Screensaver Persistence

### PR DESCRIPTION
**Pull request by h00die**
Adds a windows user level persistence.  The screensaver is overwritten with an arbitrary executable saved as an scr file.  This is a known technique documented in MITRE's ATT&CK framework and was used in a malware back in 2017,

## Verification

- [x] Start `msfconsole`
- [x] get a shell on a box
- [x] Do: ```use windows/local/user_screensaver_persistence```
- [x] Do: ```set writabledir```
- [x] Do: ```run```
- [x] Create a ```exploit/multi/handler``` with the correct payload and options
- [x] **verify** after being patient (1min by default) you get a user level shell.

### CODE REVIEW

File: documentation/modules/exploit/windows/local/user_screensaver_persistence.rb [hunk 1]
```diff
@@ -0,0 +1,134 @@
+## Vulnerable Application
```
Comment (bcoles) [inline documentation/modules/exploit/windows/local/user_screensaver_persistence.rb]: Rename `documentation/modules/exploit/windows/local/user_screensaver_persistence.rb` to `.md`

**Review by bcoles — DISMISSED** [review]

**Comment by bcoles** [comment]
Seems legit. This was on my todo list.

Is there a way to fork the original screensaver so as to have the original screensaver still execute, rather than leaving the screen unlocked?

Also of interest, perhaps as a candidate for a separate `local` module, is detection and exploitation of writable screensaver paths as a privesc vector. I've occasionally seen custom company-branded screensavers distributed in a directory writable by All Users (ie, `ProgramData`). For this reason, CIS / STIG guidelines recommend using `scrnsave.scr`, which defaults to the system path, and should never be writable by low privilege users.

File: documentation/modules/exploit/windows/local/user_screensaver_persistence.rb (hunk 1 above)
Comment (h00die) [reply documentation/modules/exploit/windows/local/user_screensaver_persistence.rb]: wow. i think i committed a cardinal sin for the docs guy

**Comment by h00die** [comment]
The *ugly* and theoretical way would be to write a batch/ps script to kick off the payload, then the original screensaver.  Theoretical because it's untested.

**Comment by sempervictus** [comment]
Batch file, or maybe PSH, but I like the idea of appearing untouched while persisting in plain sight. Should come in very handy, thank you.
/me adds to merge list

**Comment by h00die** [comment]
I'll try to investigate a strategy over the weekend.  Trying to finish up the first pass of ubiquity libs first

**Comment by h00die** [comment]
Some notes on this:
1. the file *must* have a `.scr` extension and be a native win32 executable.  Setting it as a `.bat` gets silently rejected.  Simply naming the `.bat` as `.scr` silently fails with 'not a valid win32 application' error.
2. The registry value seems to only like a file name with path, no parameters can be passed
3. I removed the ability to change the screen saver timeout, and 'is secure'.  Keep those what they were, better to be more stealth here.

A few options here to get a win32 exec to run our payload and then the screensaver:
1. Convert bat/ps1 to an exe on the fly (looks possible but seems annoyingly dependent on other scripts etc)
2. A stub exe which could read a pre-determined or coded-in file name and run the commands inside of it.
3. Load `windows/exec` payload and write it as an exe, which then calls the batch file to run the payload.

I went with option 3, keeps things from being hard coded and has no extra dependencies to be brought into msf.  However, I was unsuccessful in getting the `scr` to run.  The batch file and payload are good (even though the screensaver pops up an options screen), but the new generated `windows/exec` payload is not running the batch file.  Any ideas on where I may be going wrong?

**Comment by h00die** [comment]
as an update:
1. Re-enabled setting timeout.  `0` default keeps it unchanged.  Emailed ATT&CK, the variable is `screensavetimeout` not `screensaveRtimeout`.
2. Scr stub is now working.
3. To execute the screensaver you pass `/s`, so that is now working.


On final testing now, seems like the registry value to set the screensaver timeout is being set, yet it isn't taking effect.

**Comment by h00die** [comment]
I believe this module is ready for testing.

**Comment by bwatters-r7** [comment]
So.... this is interesting.  When I run this, I get a shell, but this is the outcome on the target.

image

Also, even if you click on the 'OK' button, the entire VM freezes while the session is running:
image

**Comment by bwatters-r7** [comment]
Interesting things I am noticing.....
I adapted the bat file to try and see why it can't find the Mystify screensave, and came across this:
When I am in a regular cmd window:
image
When I am in a cmd window spawned by the cmd_exec payload:
image

That got me thinking that maybe this is dealing with filesystem redirection with syswow.   I changed the scr payload type to windows/x64/exec and tried again.  It works.

Best guess is that the scr payload's architecture must match the original session architecture so that the filesystem locations (pulled by the original session) are the same as the ones run by the scr payload to avoid syswow redirection.

**Comment by bwatters-r7** [comment]
Unfortunately, I spoke too soon.  The screensaver comes on and you get a session, but the target computer is still locked out until the session exits.

**Comment by elibr1212** [comment]
is not work man ... 
1. when im restart the pc is not get shell
2.Even I re-run the software is not Change anything in the registry value

**Comment by h00die** [comment]
Please provide more info like what type of shell you had, what operating system your target is, what user you were doing it against, console output.

Also there are some needed changes, @bwatters-r7  and I have chat and he had some changes queues up on bugs he had found

**Comment by bwatters-r7** [comment]
For later...
https://pentestlab.blog/2019/10/09/persistence-screensaver/

**Comment by bwatters-r7** [comment]
Uh.... if we're keeping notes, it turns out that using `exitfunc` as `thread` appears to lock the computer at a blank blue screen permanently.  Not entirely surprised by that, but it should probably be said somewhere....
3x9ARXY

**Comment by bwatters-r7** [comment]
I revisited this today.  I tried powershell payloads and web-delivered powershell payloads similar to the posted article.  I hard-coded the command into the bash script and replaced the entry in the registry with the powershell command as well.  Both resulted in getting a shell that blocked the user.  The shell continued even when the user returned, leaving a blank screen until I exited the session. 😢

**Comment by h00die** [comment]
The technique looks pretty similar: https://github.com/samratashok/nishang/blob/master/Backdoors/Add-ScrnSaveBackdoor.ps1#L107
I'll test more later, but prob some very slight nuance that's hanging this up.

**Comment by h00die** [comment]
So I missed a few lines later. It looks like they aren't actually setting the payload as the screensaver, but setting a debugger to the screensaver to run then launch
https://github.com/samratashok/nishang/blob/master/Backdoors/Add-ScrnSaveBackdoor.ps1#L113
Which really is what you've already written @bwatters-r7 
https://github.com/rapid7/metasploit-framework/pull/12375/files#diff-86d755b16dc0a2bcef36f538d9e9d792R79

I'm certainly not opposed to re-writing in this method.  Untested, it would get through some of the issues we're in right now.

**Comment by bwatters-r7** [comment]
Yeah; I love the simplicity behind this PR, though.  The other is cool, but it (1) requires to know what binary someone else likes running, and (2) requires superuser permissions.  This one is great because anyone can set their screensaver, and if you suddenly end up with one turned on, you blame IT.

**Comment by h00die** [comment]
Uploading a WIP here. Lots of commented out code to document whats been tried or not.
On my 2012 box, I can't seem to get any powershell payloads to work at all. I've been trying the `windows\powershell_reverse_tcp` payload and just never seem to get a callback. That being said, Windows isn't my specialty, so it may be something obvious that i'm missing.

**Comment by bwatters-r7** [comment]
I tried again to get this to work in a non-blocking fashion, but I could not.  Attic'ing for now.

**Pull request closed without merging**